subprojects/version-control/src/main/java/org/gradle/vcs/internal/DefaultVersionControlSystemFactory.java

```diff
@@ -16,17 +16,66 @@
 
 package org.gradle.vcs.internal;
 
+import org.gradle.cache.CacheRepository;
+import org.gradle.cache.FileLockManager;
+import org.gradle.cache.PersistentCache;
 import org.gradle.vcs.VersionControlSpec;
 import org.gradle.vcs.VersionControlSystem;
+import org.gradle.vcs.VersionRef;
+import org.gradle.vcs.git.GitVersionControlSpec;
 import org.gradle.vcs.git.internal.GitVersionControlSystem;
 
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import static org.gradle.cache.internal.filelock.LockOptionsBuilder.mode;
+
 public class DefaultVersionControlSystemFactory implements VersionControlSystemFactory {
+    private final CacheRepository cacheRepository;
+    private final Map<Class<? extends VersionControlSpec>, VersionControlSystem> versionControlSystems;
+
+    DefaultVersionControlSystemFactory(CacheRepository cacheRepository) {
+        this.cacheRepository = cacheRepository;
+        // TODO: Move this map to a registry and inject the registry.
+        this.versionControlSystems = new HashMap<Class<? extends VersionControlSpec>, VersionControlSystem>();
+        versionControlSystems.put(DirectoryRepository.class, new SimpleVersionControlSystem());
+        versionControlSystems.put(GitVersionControlSpec.class, new GitVersionControlSystem());
+    }
+
     @Override
     public VersionControlSystem create(VersionControlSpec spec) {
-        // TODO: Register these mappings somewhere
-        if (spec instanceof DirectoryRepository) {
-            return new SimpleVersionControlSystem();
+        return new ThreadSafeVersionControlSystem(versionControlSystems.get(spec.getClass()), cacheRepository);
+    }
+
+    private static final class ThreadSafeVersionControlSystem implements VersionControlSystem {
+        private final VersionControlSystem delegate;
+        private final CacheRepository cacheRepository;
+
+        private ThreadSafeVersionControlSystem(VersionControlSystem delegate, CacheRepository cacheRepository) {
+            this.delegate = delegate;
+            this.cacheRepository = cacheRepository;
+        }
+
+        @Override
+        public Set<VersionRef> getAvailableVersions(VersionControlSpec spec) {
+            return delegate.getAvailableVersions(spec);
+        }
+
+        @Override
+        public void populate(final File workingDir, final VersionRef ref, final VersionControlSpec spec) {
+            PersistentCache cache = cacheRepository
+                .cache(workingDir)
+                .withLockOptions(mode(FileLockManager.LockMode.Exclusive))
```

Comment:
This should be `None` vs `Exclusive`.  It's not really wrong to use `Exclusive`.  It just means that we'll lock the cache immediately when we open it.  Since we immediately try to use the cache here and there aren't many other users of this cache, it's not very different to use `None`.  I think this will help if we need to lock other operations as well.  In that case, we might create the `PersistentCache` in the constructor of `ThreadSafeVersionControlSystem` and just call `useCache()` elsewhere.

Reply:
`None` says, `No synchronisation is done.`
https://github.com/gradle/gradle/blob/4e5372aaf8828749bdcc4a71dd20a98f2a172c90/subprojects/persistent-cache/src/main/java/org/gradle/cache/FileLockManager.java#L60-L63

I don't understand what it means to lock without synchronization. Isn't the whole point of a lock, to synchronize access?

Reply:
It's confusingly named. `withLockOptions` sets the locking behavior, but the different types affect its initial locking state.
None - do not lock initially, but lock on demand
Shared - do not lock initially, but allow multiple things to share a lock on a read-only resource
Exclusive - lock on open and hold lock until cache is closed.

See:
https://github.com/gradle/gradle/blob/92fb92fa80fb56a96461224323152c950d73a511/subprojects/persistent-cache/src/main/java/org/gradle/cache/CacheBuilder.java#L94-L100

I think it goes something like:
None is usually what you want if you're reading/writing to a shared cache.
Shared is usually what you want if multiple things are reading a resource that only requires synchronization when it's created
Exclusive is when you want to read/write to something and you aren't accessing it over and over.

Reply:
Fixed.